### ISSUE TITLE
[RFC] Make more operations inplace (GELU, BatchNorm, LayerNorm)

### ISSUE THREAD

**Issue by andoorve** [pytorch/pytorch#92884]
### 🚀 The feature, motivation and pitch

# Make more operations inplace (GELU, BatchNorm, LayerNorm)

## **Summary**
Hi PyTorch team,

We would like to enable users to make the following operations inplace: LayerNorm, BatchNorm and GELU.

## **Motivation**
In-place operations can lower the amount of memory used by operators in specific cases with minimal overhead
compared to other methods of reducing the memory footprint of models such as activation checkpointing, this
can then lead to increased performance in the form of throughput gains through batch size improvements. In-place
GELU and In-place LayerNorm combined for example, leads to an activation memory footprint savings of over
10% for Huggingface BERT at a sequence length of 512.

These benefits and prior art are described in the papers listed in the prior arts. Currently, in the
github repositories associated with these papers these In-place versions of these operators are implemented as 
PyTorch modules that must be imported.

The value add of this feature would be allowing users to take advantage of these operators directly from PyTorch. 
This would have no impact to other users of the library.

## **Proposed Implementation**
Our proposed implementation is to add an inplace flag to existing modules, which can then be set to True to enable 
the in-place version, such as with dropout for example.

`torch.nn.Dropout1d(p=0.5, inplace=False)`

The existing LayerNorm module, for example, would change from

`torch.nn.LayerNorm(normalized_shape, eps=1e-05, elementwise_affine=True, device=None, dtype=None)`
to 
`torch.nn.LayerNorm(normalized_shape, eps=1e-05, elementwise_affine=True, inplace=False, device=None, dtype=None)`.

Internally, this can then be handled as other inplace operators with a custom implementation in ATen, and adding 
the appropriate flags in the Autograd engine. Suggestions on how to do this would be welcome.

## **Metrics**
The main metric to evaluate whether this change is successful is to evaluate the memory usage of a model with
and without this change.

## **Drawbacks**
There are little to no drawbacks considering that the implementation cost is low, it does not introduce any breaking changes,
and this is not a torch wide change that would affect all users.

However, we note the In-place GELU is lossy, and this needs to be communicated/handled correctly.

## **Alternatives**
An alternative design would be to add these In-place versions as separate operators. For example, add an
`InPlaceLayerNorm` layer as its own separate module.

## **Prior Art**
These ideas are introduced in the following papers:

[In-Place Activated BatchNorm for Memory-Optimized Training of DNNs (CVPR '18)](https://openaccess.thecvf.com/content_cvpr_2018/papers/Bulo_In-Place_Activated_BatchNorm_CVPR_2018_paper.pdf)
Github repository: https://github.com/mapillary/inplace_abn

[Tempo: Accelerating Transformer-Based Model Training through Memory Footprint Reduction (NeurIPS '22)](https://openreview.net/forum?id=xqyEG7EhTZ)
Github repository: https://github.com/UofT-EcoSystem/Tempo

Example usage of Tempo:

```
from tempo.inplace_gelu import InplaceGelu
from tempo.inplace_layernorm import InplaceLayerNorm
from tempo.combined import Combined

# Original LayerNorm in Transformer model
# self.LayerNorm = nn.LayerNorm(config.hidden_size, eps=1e-12)
# Replace it with Tempo's InplaceLayerNorm
self.LayerNorm = InplaceLayerNorm(config.hidden_size, eps=1e-12)
```

## **Unresolved questions**
* How to deal with the lossy nature of GELU
* Exact implementation details

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/26288 https://github.com/pytorch/pytorch/issues/23756

**Comment by albanD** [comment]
Hey!
I think one of the reason for not having these is that these functions are used during training and we do need the input value to be able to compute the backward. So making them inplace is not very interesting because we would still need to make a copy of the input so that the backward can run properly.

**Comment by andoorve** [comment]
Hi @albanD , just to clarify the RFC is in fact *for* training! While you are correct that normally we need the input for the backward, the contribution that is made in the above papers and the point of our RFC is that you don’t actually need to stash the input, you can do the backward using the output alone. So, we can discard the input and save memory during training!

**Comment by vadimkantorov** [comment]
@albanD as you know, these are still useful for implementing fused inplace reversible nets :)

**Comment by albanD** [comment]
Sure, but in both cases, you will need a higher level system that is able to figure out that these are ok and use them.

In any case, if you need those, we would be happy to review a PR adding them (including properly sharing code with existing impl, doing proper autograd impl) but I don't think anyone on our end will work on this.

**Comment by andoorve** [comment]
@albanD that sounds good, thanks! Just a couple of questions/comments before implementing. 

I don’t think the higher level system is necessary, there’s no situation where these layers wouldn’t be ok to use and the user can choose whether to use them or not based on if it’s beneficial to their use case. 

Second, which of the implementation alternatives above would you suggest? Do you recommend the implementation making these layers completely separate, (i.e. add a new Module called `InPlaceLayerNorm`) or add them as flags to the existing layers (i.e. add an `inplace=True` flag)?

**Comment by albanD** [comment]
> I don’t think the higher level system is necessary, there’s no situation where these layers wouldn’t be ok to use and the user can choose whether to use them or not based on if it’s beneficial to their use case.

yes agreed.

> Second, which of the implementation alternatives above would you suggest? Do you recommend the implementation making these layers completely separate, (i.e. add a new Module called InPlaceLayerNorm) or add them as flags to the existing layers (i.e. add an inplace=True flag)?

The flag `inplace=True` is good for the nn.Module level.
For the nn.functional API called by the Module, an inplace: bool flag is good as well.
For the torch. API level that is called after that, this one comes directly from the c++ impl (defined in native_functions.yaml). There we need two functions: `layer_norm` (out of place) and `layer_norm_` (inplace). Each should get its own autograd formula, and the impl should route to the same thing in the end.

**Comment by vadimkantorov** [comment]
For reference, here're two PRs that were not completed:
- https://github.com/pytorch/pytorch/pull/61821
- https://github.com/pytorch/pytorch/pull/76149

**Comment by cpuhrsch** [comment]
@vadimkantorov - with torch.compile we're able to fuse many pointwise and reduction operations. Wouldn't this address your needs as well?

**Comment by vadimkantorov** [comment]
@cpuhrsch we discussed it recently in https://github.com/pytorch/pytorch/issues/92927, and it seems that for now torch.compile doesn't produce inplace code

**Comment by Chillee** [comment]
@vadimkantorov I think torch.compile does something better than producing in-place code in your cases.

**Comment by vadimkantorov** [comment]
Probably both fruits can be had which will make it even better! :)
